Fix/6569 fix partitions script in camdecmps schema

ayang2017 opened the pull request.

ayang2017 committed:
Add files via upload

ayang2017 committed:
fix/6569-Fix Partitions Script in CAMDECMPS Schema

ayang2017 commented:
Need know the Scripts for drop empty partition tables-- there are 23 tables plus 26 for emission_view_XXXX. Should save to one script or multiple ones as for each main tables?

djw4erg reviewed (changes requested):
## Requested Changes

### Main Data Tables

1. The purpose for having argument **executeFlag ** is to allow the SP to perform a dry run.  The dry run should output the SQL that would execute on an actual run.  The way this SP is setup, it displays a message about the dry run, but does not otherwise display the SQL.
2.  **DAILY_BACKSTOP** is missing from the tableList3 list. 
3. The earliest reporting period is implemented as the earliest Q1 for the earliest year.  The earliest reporting period for the three **NSPS4T** tables is **2017 Q4**, and for the **Daily Fuel** table is **2011 Q3**.

### Emission View Tables

1. The partitions are limited to 2015 and after.  That is the condition for the MATS tables (emission_view_mats%).  All other Emission View tables should have partitions for 2003 and after.

## Potential New Code Structure (but not necessary):

```
do $$
declare
    -- Arguements
    vYear smallint = 2023;
    vExecuteFlag boolean = false;
    
    -- Table List Array Variable
    vTables text[] = ARRAY
                     [
                        [ 'daily_backstop',             '125', '2024', '1' ],
                        [ 'daily_calibration',          '41',  '2003', '1' ],
                        [ 'daily_emission',             '41',  '2003', '1' ],
                        [ 'daily_fuel',                 '75',  '2011', '3' ],
                        [ 'daily_test_summary',         '41',  '2003', '1' ],
                        [ 'derived_hrly_value',         '41',  '2003', '1' ],
                        [ 'emission_evaluation',        '41',  '2003', '1' ],
                        [ 'hrly_fuel_flow',             '41',  '2003', '1' ],
                        [ 'hrly_gas_flow_meter',        '89',  '2015', '1' ],
                        [ 'hrly_op_data',               '41',  '2003', '1' ],
                        [ 'hrly_param_fuel_flow',       '41',  '2003', '1' ],
                        [ 'long_term_fuel_flow',        '41',  '2003', '1' ],
                        [ 'mats_derived_hrly_value',    '89',  '2015', '1' ],
                        [ 'mats_monitor_hrly_value',    '89',  '2015', '1' ],
                        [ 'monitor_hrly_value',         '41',  '2003', '1' ],
                        [ 'nsps4t_annual',              '100', '2017', '4' ],
                        [ 'nsps4t_compliance_period',   '100', '2017', '4' ],
                        [ 'nsps4t_summary',             '100', '2017', '4' ],
                        [ 'sampling_train',             '89',  '2015', '1' ],
                        [ 'sorbent_trap',               '89',  '2015', '1' ],
                        [ 'summary_value',              '41',  '2003', '1' ],
                        [ 'weekly_system_integrity',    '89',  '2015', '1' ],
                        [ 'weekly_test_summary',        '89',  '2015', '1' ]
                     ];
    
    -- Other Variables
    vTable text[];
    vTableName text;
    vEarliestRptPeriodId smallint;
    vEarliestYear smallint;
    vEarliestQuarter smallint;
    vSqlStatement text;
    prd record;
    tbl record;
begin
    
    -- Handle Emission Data Tables
    foreach vTable slice 1 in array vTables
    loop
        vTableName = vTable[1];
        vEarliestRptPeriodId = vTable[2]::smallint;
        vEarliestYear = vTable[3]::smallint;
        vEarliestQuarter = vTable[4]::smallint;
        
        for prd in
        (
            select  rpt_period_id,
                    calendar_year,
                    quarter
              from  camdecmpsmd.REPORTING_PERIOD
             where  calendar_year = vYear
               and  vYear >= 2003
               and  (
                        calendar_year > vEarliestYear
                        or
                        calendar_year = vEarliestYear and quarter >= vEarliestQuarter
                    )
        )
        loop
            
            vSqlStatement = format('CREATE TABLE IF NOT EXISTS camdecmps.%s_%s_q%s PARTITION OF camdecmps.%s FOR VALUES FROM (%s) TO (%s);    ', 
                                    vTableName, prd.calendar_year, prd.quarter, vTableName, prd.rpt_period_id, prd.rpt_period_id + 1 );
            
            if vExecuteFlag then
                raise notice '%', vSqlStatement;
                raise notice 'UNCOMMENT EXECUTION';
                --execute vSqlStatement;
            else
                raise notice 'Dry Run includes "%".', vSqlStatement;
            end if;
            
        end loop;
        
    end loop;
    
    
    -- Handle Emission View Tables
    for tbl in
    (
        select  table_name
          from  information_schema.TABLES
         where  table_schema = 'camdecmps'
           and  table_type = 'BASE TABLE'
           and  table_name like 'emission_view_%'
           and  table_name not like 'emission_view_%q%'
    )
    loop
        
        for prd in
        (
            select  rpt_period_id,
                    calendar_year,
                    quarter
              from  camdecmpsmd.REPORTING_PERIOD
             where  calendar_year = vYear
               and  vYear >= 2003
               and  (
                        tbl.table_name not like 'emission_view_mats%'
                        or
                        vYear >= 2015
                    )
        )
        loop
            
            vSqlStatement = format('CREATE TABLE IF NOT EXISTS camdecmps.%s_%s_q%s PARTITION OF camdecmps.%s FOR VALUES FROM (%s) TO (%s);    ', 
                                    tbl.table_name, prd.calendar_year, prd.quarter, tbl.table_name, prd.rpt_period_id, prd.rpt_period_id + 1 );
            
            if vExecuteFlag then
                raise notice '%', vSqlStatement;
                raise notice 'UNCOMMENT EXECUTION';
                --execute vSqlStatement;
            else
                raise notice 'Dry Run includes "%".', vSqlStatement;
            end if;
            
        end loop;
        
    end loop;

end
$$;
```

ayang2017 committed:
Modified upon Dwayne's review

Included the detailed information for related partition table in the ARRAY make perfect sense.

ayang2017 committed:
Add files via upload

ayang2017 commented:
It make perfect sense to included related partition info in the ARRAY. Updated the SP based on Dwayne's suggestion.

ayang2017 committed:
Two Scripts for drop empty partition tables

One for emission tables(23 main tables), one for emission_view tables (26 main tables).

ayang2017 commented:
Loaded two scripts for drop empty partition tables at deployment folder.

djw4erg reviewed (changes requested):
1. Because this is not a stored procedure instead of a script, it should move from the Partitions folder for CAMDECMPS to the Procedures folder.
   - Make sure that @mark-hayward-erg  agrees.
   - I can help with doing a Git move to try to retain history.
2. The actual execute of the statements when executeflag is True is still commented out, and the Raise Notice indicating that the execute needs to be uncommented still exists.

mark-hayward-erg commented:
> 1. Because this is not a stored procedure instead of a script, it should move from the Partitions folder for CAMDECMPS to the Procedures folder.
>    
>    * Make sure that @mark-hayward-erg  agrees.
>    * I can help with doing a Git move to try to retain history.
> 2. The actual execute of the statements when executeflag is True is still commented out, and the Raise Notice indicating that the execute needs to be uncommented still exists.

I agree that the new procedure should be added to the procedures folder.

However, there still should be a create-partitions.sql script in the camdecmps/partitions folder that calls the new procedure.  This way, the existing db build sequence can function the same way it does now.

ayang2017 committed:
Delete camdecmps/partitions/create_partitions.sql

Name not match previous one.

ayang2017 committed:
Add files via upload

Updated SP.

ayang2017 committed:
Add files via upload

Two scripts for detach and drop empty partition tables.

ayang2017 commented:
Updated the SP "create-partitions", saved at "partition" folder. the two scripts for detach and drop partition tables were saved at "deployments" folder, now added to "procedure" folder.

djw4erg commented:
> 
> I agree that the new procedure should be added to the procedures folder.
> 
> However, there still should be a create-partitions.sql script in the camdecmps/partitions folder that calls the new procedure. This way, the existing db build sequence can function the same way it does now.

Ah yes.  That makes sense about retaining an updated version of the script that calls the SP.

djw4erg reviewed (changes requested):
Anna, I believe the following file moves need to happen in GitHub.

1. Copy **create_partitions.sql (procedure)** to the camdecmps/procedures folder.
2. Change **create_partitions.sql (script)** to call the SP for every year from 2003 through the current year and call with **executeFlag** set to true.  (This was Mark's point)
3. Delete **detach_drop_emissions.sql** to the camdecmps/partitions folder, but should be in the Deployments folder.
4. Delete **detach_drop_emissions_view.sql** to the camdecmps/partitions folder, but should be in the Deployments folder.

**Please let me know if you need help with the Git Moves and the Revert.**

ayang2017 committed:
Add files via upload

ayang2017 committed:
Delete camdecmps/procedures/Detach_drop_emissions.sql

ayang2017 committed:
Delete camdecmps/procedures/Detach_drop_emissions_view.sql

ayang2017 committed:
Delete deployments/ecmps/release-v2.0-fix/Sprint17/6569/create_partitions.sql

ayang2017 committed:
Add files via upload

The script is calling the SP for all years after 2003 to create partition tables.

ayang2017 commented:
Update--finished:
1. Copy **create_partitions.sql (procedure)** to the camdecmps/procedures folder.
2. Change **create_partitions.sql (script)** to call the SP for every year from 2003 through the current year and call with **executeFlag** set to true.  
3. Delete **detach_drop_emissions.sql** to the camdecmps/partitions folder, but should be in the Deployments folder.
4. Delete **detach_drop_emissions_view.sql** to the camdecmps/partitions folder, but should be in the Deployments folder.

djw4erg commented on camdecmps/partitions/create-partitions.sql:
```diff
@@ -1,133 +1,8 @@
+DECLARE vYear smallint;
+BEGIN 
+  SELECT EXTRACT('year' FROM CURRENT_DATE) into vYear;  
+  for counter in 2024..vYear loop
```
Based on the discussions with @mark-hayward-erg, I believe the begin year should be 2003.  This script is meant to run once.

The create-partitions.sql script in the Procedures folder needs to be added to the Deployment folder.

djw4erg reviewed (changes requested):
The create-partitions.sql scripts in both the Procedures and Partitions folders need to be added to the Deployment folder.  You will need to append "_procedure" and "_script" to their names.

djw4erg committed:
#6569 Correcting begin year in the create-partitions.sql script, and adding the create-partitions procedure script to the Deployment folder.

mark-hayward-erg commented on camdecmps/partitions/create-partitions.sql:
```diff
@@ -1,77 +1,8 @@
+BEGIN 
+  SELECT EXTRACT('year' FROM CURRENT_DATE) into vYear;  
+  for counter in 2003..vYear loop
+       call camdecmps.create-partitions(counter, 'true');
```
@ayang2017 the procedure name being called is create_partitions, not create-partitions.  Also, the file names for the create_partitions scripts should match the procedure name (i.e., with "_" instead of "-").

mark-hayward-erg commented on camdecmps/procedures/create-partitions.sql:
The file name for the create_partitions script should match the procedure name (i.e., with "_" instead of "-").

mark-hayward-erg commented on deployments/ecmps/release-v2.0-fix/Sprint17/6569/create-partitions.sql:
The file name for the create_partitions script should match the procedure name (i.e., with "_" instead of "-").

ayang2017 replied:
Name correction at 3 locations:
/partitions, /procedures, /deployment
Thanks,

ayang2017 committed:
correct file name

ayang2017 committed:
Delete camdecmps/partitions/create-partitions.sql

ayang2017 committed:
Update and rename create-partitions.sql to create_partitions.sql

ayang2017 committed:
Rename create-partitions.sql to create_partitions.sql

mark-hayward-erg commented on camdecmps/partitions/create_partitions.sql:
```diff
@@ -0,0 +1,8 @@
+BEGIN 
+  SELECT EXTRACT('year' FROM CURRENT_DATE) into vYear;  
+  for counter in 2003..vYear loop
+       call camdecmps.create-partitions(counter, 'true');
```
This still needs to be updated to call the procedure as it is define ("_" instead of "-").

ayang2017 replied:
Updated! Thanks!!

ayang2017 committed:
Update create_partitions.sql
